lib/rules/no-duplicate-id.js

```diff
@@ -0,0 +1,139 @@
+      }
+    }
+
+    // Resolve ConcatStatement parts values to StringLiteral where possible
+    function getJoinedConcatParts(node, scope) {
+      return node.value.parts.map((part) => getPartValue(part, scope)).join('');
+    }
+
+    // Store the idValues collected; reference to look for duplicates
+    let idValueSet = new Set();
+
+    return {
+      AttrNode(node) {
+        // Only check relevant nodes
+        if (!isValidIdAttrNode(node)) {
+          return;
+        }
+
+        let idValue;
+
+        // Check idValues that resolve to a StringLiteral
+
+        // TextNode: unwrap
+        // ex. id="id-value" becomes "id-value"
+        if (node.value.type === 'TextNode') {
+          idValue = node.value.chars;
+        }
+
+        // MustacheStatement: try to resolve
+        // ex. id={{"id-value"}} becomes "id-value"
+        // ex. id={{idValue}} becomes "{{idValue}}"
+        if (node.value.type === 'MustacheStatement') {
+          idValue = getMustacheValue(node.value, this);
+        }
+
+        // ConcatStatement: try to resolve parts to StringLiteral where possible
+        // ex. id="id-{{"value"}}" becomes "id-value"
+        // ex. id="id-{{value}}-{{"number"}}" becomes "id-{{value}}-number"
+        if (node.value.type === 'ConcatStatement') {
+          idValue = getJoinedConcatParts(node, this);
+        }
+
+        // If idValue is not assigned by this point, use the raw source
+        if (!idValue) {
+          idValue = this.sourceForNode(node.value);
+        }
+
+        if (idValueSet.has(idValue)) {
+          this.logNode({ node, message: ERROR_MESSAGE });
+        } else {
+          idValueSet.add(idValue);
+        }
+      },
+
+      // BlockStatements store attributes in a hash pair, so we have to check these in a different way
+      BlockStatement(node) {
+        let refPair = { key: 'elementId', value: { type: 'StringLiteral' } };
+        let elementIdHashArg = node.hash.pairs.find((testPair) =>
+          NodeMatcher.match(testPair, refPair)
+        );
+        if (!elementIdHashArg) {
+          return;
+        }
+        let idValue = elementIdHashArg.value.value;
+
+        if (idValueSet.has(idValue)) {
+          this.logNode({ node, message: ERROR_MESSAGE });
+        } else {
+          idValueSet.add(idValue);
+        }
+      },
+
+      // MustacheStatements also store attributes in a hash pair but they do not use `elementId` like blockStatements, they use `id` as the attribute name
+      MustacheStatement(node) {
+        let refPair = { key: 'id', value: { type: 'StringLiteral' } };
+        let idHashArg = node.hash.pairs.find((testPair) => NodeMatcher.match(testPair, refPair));
+
+        if (!idHashArg) {
+          return;
+        }
+
+        let idValue = idHashArg.value.value;
+
+        if (idValueSet.has(idValue)) {
+          this.logNode({ node, message: ERROR_MESSAGE });
+        } else {
+          idValueSet.add(idValue);
+        }
```

Comment:
This seems like this is duplicated between `BlockStatement` and here, can we share the code?